src/xenstore.c

```diff
@@ -536,3 +590,406 @@ xenstore_deinit(void)
   xs_daemon_close(xs_handle);
   xs_handle = NULL;
 }
+
+void
+xsdev_watch_deinit(void)
+{
+  if (watch_path) {
+    xs_unwatch(xs_handle, watch_path, watch_token);
+    free(watch_path);
+    watch_path = NULL;
+  }
+}
+
+/* Returns true (1) on success, false (0) on failure. */
+int
+xsdev_watch_init(void)
+{
+  if (usb_backend_domid > 0) {
+    watch_path = xasprintf("%s/data/usb", xs_backend_path);
+    return xs_watch(xs_handle, watch_path, watch_token);
+  }
+
+  return 1;
+}
+
+void xsdev_del(device_t *dev)
+{
+  if (g_xcbus) {
+    return;
+  }
+
+  xd_log(LOG_INFO, "%s %d %d %s", __func__, dev->busid, dev->devid,
+         dev->sysname);
+
+  char *path = xasprintf("data/usb/dev%d-%d", dev->busid, dev->devid);
+  if (path == NULL) {
+    xd_log(LOG_ERR, "%s: xasprintf path failed", __func__);
+    return;
+  }
+
+  char *vusb_watch = xasprintf("%s/assign", path);
+  char *vusb_token = xasprintf(ASSIGN_PREFIX"%x %x", dev->vendorid,
+                 dev->deviceid);
+
+  xs_unwatch(xs_handle, vusb_watch, vusb_token);
+
+  free(vusb_watch);
+  free(vusb_token);
+
+  if (xs_rm(xs_handle, XBT_NULL, path) == false) {
+    xd_log(LOG_ERR, "failure xs_rm(%s)", path);
+  }
+
+  free(path);
+}
+
+void xsdev_write(device_t *dev)
+{
+  xs_transaction_t t;
+
+  if (g_xcbus) {
+    return;
+  }
+
+  xd_log(LOG_INFO, "%s %d %d %s", __func__, dev->busid, dev->devid,
+         dev->sysname);
+
+  char *path = xasprintf("data/usb/dev%d-%d", dev->busid, dev->devid);
+  if (path == NULL) {
+    xd_log(LOG_ERR, "%s: xasprintf path failed", __func__);
+    return;
+  }
+
+  xenstore_add_dir(XBT_NULL, "data/usb", my_domid, XS_PERM_NONE,
+                             0, XS_PERM_READ);
+
+  /* TODO Add transaction abort? */
+ restart:
+  t = xs_transaction_start(xs_handle);
+
+  xs_mkdir(xs_handle, t, path);
+
+#define xs_write_fmt(fmt, v) { \
```

Comment:
What was the reasoning for the #define #undef mid-function here?

Reply:
I \#undef the macro to keep it local to the function.

They are just to cut down on repetition.  Also, It lets me "stringify" the file name for writing into xenstore.